buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/ci/AffectedProjectFinder.groovy

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.gradle.plugins.ci
+
+import groovy.transform.builder.Builder
+
+import java.util.regex.Pattern
+import org.gradle.api.Project
+
+/** Determines a set of subprojects that own the 'changedPaths'. */
+class AffectedProjectFinder {
+    Project project;
+    Set<String> changedPaths;
+
+    @Builder
+    AffectedProjectFinder(Project project,
+                          Set<String> changedPaths,
+                          List<Pattern> ignorePaths) {
+        this.project = project
+        this.changedPaths = changedPaths.findAll {
+            for(def path : ignorePaths) {
+                if(it ==~ path) {
+                    return false
+                }
+            }
+            return true
+        }
+    }
+
+    Set<Project> find() {
+        Set<String> paths = changedPaths.collect()
```

Comment:
What is the purpose of this line? changedPaths is already a Set and no transform is provided to collect.